LibWeb: Support CSS style `background-clip: text`

zacbrannelly opened the pull request:
Support for the `background-clip: text` CSS property, currently only supports the CPU renderer, will setup a machine that supports the GPU renderer (I'm on a Mac M1) and implement in either this PR or a subsequent one. Also probably have to fix a bunch of linting errors since I haven't got that setup on my machine properly.

kalenikaliaksandr commented on Userland/Libraries/LibWeb/Painting/Paintable.cpp:
```diff
@@ -172,4 +172,19 @@ CSSPixelPoint Paintable::box_type_agnostic_position() const
     return position;
 }
 
+Optional<u32> Paintable::get_alpha_mask_id_from_ancestors() const
+{
+    Optional<u32> alpha_mask_id;
+    auto const containing_block = layout_node().containing_block();
+    for (auto* ancestor = containing_block; ancestor; ancestor = ancestor->containing_block()) {
+        if (ancestor->paintable() && ancestor->paintable()->fast_is<PaintableWithLines>()) {
+            auto const& ancestor_with_lines = static_cast<PaintableWithLines const&>(*ancestor->paintable());
+            alpha_mask_id = ancestor_with_lines.get_alpha_mask_id();
+            if (alpha_mask_id.has_value()) break;
+        }
+    }
+
+    return alpha_mask_id;
+}
```
traversing the containing block chain is something we try to avoid because:

- it makes painting commands recording noticeably slower and very visible in profiles on any large page
- adds yet another dependency of paintable tree on layout tree

so we are in the process of getting rid of this pattern.
instead, could we assign each paintable that has to be affected by mask corresponding mask id on pre-paint phase, like we do that for overflow clipping and scrolling offsets?
you could check how it works by looking here:
https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/Painting/ViewportPaintable.cpp#L91
https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/Painting/ViewportPaintable.cpp#L136

zacbrannelly replied:
This bit of feedback should be addressed now :)

kalenikaliaksandr commented:
also some ref-tests would be much appreciated. you can find examples of the once we already have here:
https://github.com/SerenityOS/serenity/tree/master/Tests/LibWeb/Ref

zacbrannelly commented on Userland/Libraries/LibWeb/Painting/CommandExecutorCPU.h:
```diff
@@ -67,10 +72,34 @@ class CommandExecutorCPU : public CommandExecutor {
+    [[nodiscard]] Gfx::Painter& target_painter() { return *stacking_contexts.last().painter; }
 
     Vector<StackingContext> stacking_contexts;
+    Vector<OwnPtr<ForegroundTextAlphaMask>> m_foreground_text_alpha_masks;
+    Vector<u32> m_alpha_mask_stack;
```
Unfortunately not, at least with the approach I've gone with, the way it works is when an alpha mask is set active all non-glyph commands paint to a separate background bitmap, once all children have been painted we apply the alpha mask to the background bitmap and then paint it to the final target canvas. If all background commands  were not aware of the masking then they would paint on the target bitmap, and I don't see how we can apply the alpha mask directly to the target bitmap as the masked pixels need to be replaced with the colour of the content behind the background, this information is only known before the background is painted.

Let me know if you can think of a better way to do this.

kalenikaliaksandr replied:
> Unfortunately not, at least with the approach I've gone with, the way it works is when an alpha mask is set active all non-glyph commands paint to a separate background bitmap, once all children have been painted we apply the alpha mask to the background bitmap and then paint it to the final target canvas.

what if instead of waiting for all children to be painted, a mask were applied to each DrawGlyphRun command separately?

> If all background commands were not aware of the masking then they would paint on the target bitmap, and I don't see how we can apply the alpha mask directly to the target bitmap as the masked pixels need to be replaced with the colour of the content behind the background, this information is only known before the background is painted.

what do you mean by background commands?

zacbrannelly replied:
> what if instead of waiting for all children to be painted, a mask were applied to each DrawGlyphRun command separately?

The mask cannot be applied when drawing the glyphs because that would mask out the background layer for the subsequent glyphs. We need to wait for all text to be painted to the alpha mask so we can then clip out the background only where there is no text.

> what do you mean by background commands?

I mean the paint commands that are not text related that contribute to the background layer that we are clipping. Commands like `fill_rect` / `draw_scaled_bitmap`.

kalenikaliaksandr replied:
> I mean the paint commands that are not text related that contribute to the background layer that we are clipping.

what happens if glyph painting commands are interleaved with "background painting commands"? I build this branch to check but it crashes on this page, for example:
```html
<!DOCTYPE html>
<style>
.gradient-text {
    font-size: 60px;
    font-weight: bold;
    color: transparent;
    background-image: linear-gradient(45deg, #ff69b4, #8a2be2);
    background-clip: text;
    display: inline-block;
}
</style>
<div class="gradient-text">Gradient Text<div style="display: inline-block; opacity: 0.5">hmmm</div></div>
```

zacbrannelly replied:
Ahhh good catch! I clearly need to do more testing and read up on the codebase / specs more 😅 (this is my first attempt at contributing and still learning), removing line 502 in `CommandExecutorCPU.cpp` fixes the crash, but it's definitely not matching the behaviour of Chrome, plus text highlighting is a bit broken too, so will look into it. Cheers :D

This is the code I've been using for testing so far:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>

  <style>
    p, .container {
      border: 0.8em darkviolet;
      border-style: dotted double;
      margin: 1em 0;
      padding: 1.4em;
      background: linear-gradient(60deg, red, yellow, red, yellow, red);
      font: 900 1.2em sans-serif;
      text-decoration: underline;
    }

    .border-box {
      background-clip: border-box;
    }
    .padding-box {
      background-clip: padding-box;
    }
    .content-box {
      background-clip: content-box;
    }

    .text {
      background-clip: text;
      color: rgb(0 0 0 / 20%);
    }
  </style>
</head>
<body>
  <p class="border-box">The background extends behind the border.</p>
  <p class="padding-box">
    The background extends to the inside edge of the border.
  </p>
  <p class="content-box">
    The background extends only to the edge of the content box.
  </p>
  <div class="text container">
    The background is clipped to the foreground text.
    <span>Some other text in a sub-element</span>
  </div>
  <div class="text container">
    <div>
      <div>The is nested text that should still be clipped to the background</div>
    </div>
  </div>
</body>
</html>
```

kalenikaliaksandr commented on Userland/Libraries/LibWeb/Painting/Paintable.h:
```diff
@@ -212,6 +217,7 @@ class Paintable
     Optional<JS::GCPtr<Layout::Box const>> mutable m_containing_block;
 
     OwnPtr<StackingContext> m_stacking_context;
+    Optional<u32> m_alpha_mask_id;
```
shouldn't alpha mask id be only stored in TextPaintable?

zacbrannelly replied:
Both `PaintableBox` and `TextPaintable` need the alpha mask id, since it maps to both the alpha mask and the background canvas that `PaintableBox` needs to paint the background layer to. So I can move this to just those two classes instead of the parent if thats preferred?

Also the variable name might be a bit 💩, just haven't come up with one that clearly explains what the id maps to.

zacbrannelly commented on Userland/Libraries/LibWeb/Painting/PaintableBox.cpp:
```diff
@@ -667,6 +680,9 @@ void paint_text_fragment(PaintContext& context, Layout::TextNode const& text_nod
 
         auto text = text_node.text_for_rendering();
 
+        if (should_paint_to_alpha_mask)
+            painter.push_alpha_mask_id(foreground_text_alpha_mask_id.value());
```
Yeah we could get rid of the push/pop calls in this part of the code by saving it in the `DrawGlyphRun` command, but we still need them for the non-glyph commands that contribute to painting the background layer. Otherwise I'd need to store the alpha mask id in those commands too, which means modifying a bunch of functions, all to get a different painter, I went with the push/pop state to avoid that and support all commands straight away.

kalenikaliaksandr commented:
I took another look at the spec for the definition of how `text` clip should be handled https://drafts.csswg.org/css-backgrounds-4/#background-clip

> The background is painted within (clipped to) the intersection of the border box and the geometry of the text in the element and its in-flow and floated descendants

so what if we tried to implement in the following way?
- if element has `background-clip: text` then we calculate intersection of its border box and all glyphs from descendants. I think we can start by collecting a vector of Gfx::Path for all descendant glyphs. this way we won't need to literally calculate the intersection.
- vector with all glyph paths that are supposed to used for masking is saved in `DrawScaledImmutableBitmap` command that does background painting
- executor is updated to apply a mask of `DrawScaledImmutableBitmap` command: if vector with glyph paths that are supposed to be used for a mask is not empty then we use them to paint a separate bitmap with a mask that is applied to background's bitmap.

this way we can completely get rid of alpha mask ids, push/pop mask states, having separate painters for background and glyph commands. wouldn't it be cool to make way simpler? 🤓

zacbrannelly commented:
> I took another look at the spec for the definition of how `text` clip should be handled https://drafts.csswg.org/css-backgrounds-4/#background-clip
> 
> 
> 
> > The background is painted within (clipped to) the intersection of the border box and the geometry of the text in the element and its in-flow and floated descendants
> 
> 
> 
> so what if we tried to implement in the following way?
> 
> - if element has `background-clip: text` then we calculate intersection of its border box and all glyphs from descendants. I think we can start by collecting a vector of Gfx::Path for all descendant glyphs. this way we won't need to literally calculate the intersection.
> 
> - vector with all glyph paths that are supposed to used for masking is saved in `DrawScaledImmutableBitmap` command that does background painting
> 
> - executor is updated to apply a mask of `DrawScaledImmutableBitmap` command: if vector with glyph paths that are supposed to be used for a mask is not empty then we use them to paint a separate bitmap with a mask that is applied to background's bitmap.
> 
> 
> 
> this way we can completely get rid of alpha mask ids, push/pop mask states, having separate painters for background and glyph commands. wouldn't it be cool to make way simpler? 🤓 

Sounds much better! Only additional thing is the fill rect command will also need the glyph paths since it's not always an image that's is drawn for the background. Will give it a shot when I've got some more time, Thanks heaps! :)

zacbrannelly commented:
@kalenikaliaksandr I've implemented your suggestion with just the `FillRectWithRoundedCorners` command to test it out and running into an issue where the text isn't lining up between the background and the foreground text:
<img width="1598" alt="Screenshot 2024-03-02 at 4 04 43 pm" src="https://github.com/SerenityOS/serenity/assets/8196384/0926a5ca-15a5-4f8f-9538-242e6a351400">

Where the red colour is the background thats clipped via `Gfx::Path`'s, and the green text is the foreground text being painted on top of the background.

This is how I'm collecting the paths:
```cpp
        auto add_text_clip_path = [&](Layout::TextNode const& text_node, PaintableFragment const& fragment) {
            // Draw the glyphs to an instance of Gfx::Path.
            auto text = text_node.text_for_rendering();

            // Scale to the device pixels.
            Gfx::Path glyph_run_path;
            for (auto glyph : fragment.glyph_run()) {
                glyph.visit([&](auto& glyph) {
                    glyph.font = *glyph.font->with_size(glyph.font->point_size() * static_cast<float>(context.device_pixels_per_css_pixel()));
                    glyph.position = glyph.position.scaled(context.device_pixels_per_css_pixel());
                });

                if (glyph.has<Gfx::DrawGlyph>()) {
                    auto const& draw_glyph = glyph.get<Gfx::DrawGlyph>();

                    // Get the path for the glyph.
                    Gfx::Path glyph_path;
                    auto const& scaled_font = static_cast<Gfx::ScaledFont const&>(*draw_glyph.font);
                    auto glyph_id = scaled_font.glyph_id_for_code_point(draw_glyph.code_point);
                    scaled_font.append_glyph_path_to(glyph_path, glyph_id);

                    // Transform the path to the fragment's position.
                    auto transform = Gfx::AffineTransform {}.translate(draw_glyph.position);
                    glyph_run_path.append_path(glyph_path.copy_transformed(transform));
                }
            }

            // Calculate the baseline start position.
            auto fragment_absolute_rect = fragment.absolute_rect();
            auto fragment_absolute_device_rect = context.enclosing_device_rect(fragment_absolute_rect);
            DevicePixelPoint baseline_start { fragment_absolute_device_rect.x(), fragment_absolute_device_rect.y() + context.rounded_device_pixels(fragment.baseline()) };

            // Add the path to text_clip_paths.
            auto transform = Gfx::AffineTransform {}.translate(baseline_start.to_type<int>().to_type<float>());
            text_clip_paths.append(glyph_run_path.copy_transformed(transform));
        };
```

And this is how it is clipped in the executor:
```cpp
template<typename Callback>
void apply_clip_paths_to_painter(Gfx::IntRect const& rect, Callback callback, Vector<Gfx::Path> const& clip_paths, Gfx::Painter& target_painter)
{
    // Setup a painter for a background canvas that we will paint to first.
    auto background_canvas = Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, rect.size()).release_value_but_fixme_should_propagate_errors();
    Gfx::Painter painter(*background_canvas);

    // Offset the painter to paint in the correct location.
    painter.translate(-rect.location());

    // Paint the background canvas.
    callback(painter);

    // Apply the clip path to the target painter.
    Gfx::AntiAliasingPainter aa_painter(target_painter);
    for (auto const& clip_path : clip_paths) {
        auto fill_offset = clip_path.bounding_box().location().to_type<int>() - rect.location();
        auto paint_style = Gfx::BitmapPaintStyle::create(*background_canvas, fill_offset).release_value_but_fixme_should_propagate_errors();
        aa_painter.fill_path(clip_path, paint_style);
    }
}

CommandResult CommandExecutorCPU::fill_rect_with_rounded_corners(Gfx::IntRect const& rect, Color const& color, Gfx::AntiAliasingPainter::CornerRadius const& top_left_radius, Gfx::AntiAliasingPainter::CornerRadius const& top_right_radius, Gfx::AntiAliasingPainter::CornerRadius const& bottom_left_radius, Gfx::AntiAliasingPainter::CornerRadius const& bottom_right_radius, Vector<Gfx::Path> const& clip_paths)
{
    auto paint_op = [&](Gfx::Painter& painter) {
        Gfx::AntiAliasingPainter aa_painter(painter);
        aa_painter.fill_rect_with_rounded_corners(
            rect,
            color,
            top_left_radius,
            top_right_radius,
            bottom_right_radius,
            bottom_left_radius);
    };
    if (clip_paths.is_empty()) {
        paint_op(painter());
    } else {
        apply_clip_paths_to_painter(rect, paint_op, clip_paths, painter());
    }
    return CommandResult::Continue;
}
```

Any idea's where I'm going wrong?

zacbrannelly commented:
I think I've fixed it by doing:
```cpp
// Transform the path to the fragment's position.
auto top_left = draw_glyph.position + Gfx::FloatPoint(scaled_font.glyph_left_bearing(draw_glyph.code_point), 0);
auto glyph_position = Gfx::GlyphRasterPosition::get_nearest_fit_for(top_left);
auto transform = Gfx::AffineTransform {}.translate(glyph_position.blit_position.to_type<float>());
glyph_run_path.append_path(glyph_path.copy_transformed(transform));
```

<img width="1252" alt="Screenshot 2024-03-02 at 4 35 49 pm" src="https://github.com/SerenityOS/serenity/assets/8196384/7ad36b08-7f3d-48e6-9a19-d8e1dabbb319">

Still some of the background leaking through but not as bad as before.

kalenikaliaksandr commented:
it seems that the rendering discrepancy comes from from the difference between rasterizing Gfx::Path "as is", and how Gfx::Painter accounts for the subpixel offset (we produce couple bitmaps for each glyph and use the best one, depending on the glyph's position).

https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGfx/Painter.cpp#L1359-L1363

I think the easiest way to fix that would be to save array of (font, code_point) instead of Gfx::Path and use `Painter::draw_glyph()` so it is always the same code used for glyph painting. BUT, feel free to leave a FIXME, instead of doing that!

great progress btw! 🤓

zacbrannelly commented:
> it seems that the rendering discrepancy comes from from the difference between rasterizing Gfx::Path "as is", and how Gfx::Painter accounts for the subpixel offset (we produce couple bitmaps for each glyph and use the best one, depending on the glyph's position).
> 
> https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGfx/Painter.cpp#L1359-L1363
> 
> I think the easiest way to fix that would be to save array of (font, code_point) instead of Gfx::Path and use `Painter::draw_glyph()` so it is always the same code used for glyph painting. BUT, feel free to leave a FIXME, instead of doing that!
> 
> great progress btw! 🤓

Yeahh I've just added a FIXME for now :)
I've added a test and fixed all the linting issues, so should be good for a proper look.

The only thing I'm concerned about is the performance, when you load the test case I've added and highlight some text it is noticeably slower at repainting. Trying to do some profiling to see where its slowing down the most.

kalenikaliaksandr reviewed (approved):
thank you for working on this change! I updated the branch with a couple of little changes:
- moved css parser update into separate commit
- wrote a description for commit that implements painting part
- removed some unused variables

it looks good to merge. I profiled a couple websites that use this feature, and it seems like we have large problems to address until "background-clip: text" will become a problem 😅
